text/0003-data-frames.md

```diff
@@ -0,0 +1,252 @@
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Data frame representation:
+
+```rust
+struct DataFrame {
+    headers: Option<Vec<String>>,
+    rows: Vec<Vec<Value>>,
+    partial_frame_id: Option<Uuid>,
+    is_object: bool,
+}
+```
+
+## Tables
+
+A self-contained table would look like this:
+
+```rust
+let frame = DataFrame {
+    headers: Some(vec!["name", "age"]),
+    rows: vec![
+        vec![Value::from("Bob"), Value::from(30)],
+        vec![Value::from("Sally"), Value::from(43)]
+    ],
+    partial_frame_id: None,
+    is_object: false,
+};
+```
+
+The above code could be created using this Nu syntax:
+
+```sh
+[name: [Bob, Sally], age: [30, 43]]
+```
+
+## Lists and matrices
+
+A self-contained list would look like this:
+
+```rust
+let list = DataFrame {
+    headers: None,
+    rows: vec![
+        vec![Value::from(1), Value::from(2), Value::from(3)]
+    ],
+    partial_frame_id: None,
+    is_object: false,
+}
+```
+
+The above code could be created using this Nu syntax:
+
+```sh
+[1, 2, 3]
+```
+
+## Objects (aka hash tables)
+
+A self-contained object would look like this:
+
+```rust
+let obj = DataFrame {
+    headers: Some(vec!["name", "level"]),
+    rows: vec![
+        vec![Value::from("Thomas"), Value::from(12)]
+    ],
+    partial_frame_id: None,
+    is_object: true,
+}
+```
+
+**Note:** we use the boolean in the table rather than enumeration because all processing on the frame remains uniform regardless of if the frame is a single row with headers vs an object. This simplifies algorithms to only have to work with the data directly, and we can later represent this data and/or serialize this data in a way that maintains the user's model.
+
+The above code could be created using this Nu syntax:
+
+```sh
+[name: Thomas, level: 12]
+```
+
+## Streaming
+
+An important part of Nu is being able to work with potentially endless streams of data. As is often the case when working with external commands, there's no guarantee that the output will terminate.
+
+We need to be able to represent the results of processing a stream of unprocessed data as a stream of data frames.
+
+To be able to output a data frame as a stream, we need to know two key elements: that the data frame is incomplete as-is, and a unique identifier that allows us to later stream additional data for this frame.
+
+To accomplish this streaming, we also introduce an `EndFrame`. Frame and EndFrame work together to allow a frame to be streamed as a multi-part frame, ending once the corresponding EndFrame has been read.
+
+As an example, let's say we were processing some content and wanted to output the first row and later the second row of this table:
+
+| tag  | length |
+| ---- | ------ |
+| head | 1024   |
+| body | 8192   |
+
+To do this, we would send the two separate frames, both marked as partial frames.
```

Comment:
This could add, ", ending the stream with an `EndFrame` with the same frame id."